extension-workspace-reference.md

```diff
@@ -0,0 +1,111 @@
+interface ServerCapabilities {
+  /* ... all fields from the base ServerCapabilities ... */
+
+  /**
+   * The server provides workspace reference exporting support.
+   */
+  workspaceReferenceProvider?: boolean;
+}
+```
+
+#### Workspace References Request
+
+The workspace reference request is sent from the client to the server to export project-wide external references. The response strictly includes only the locations in the workspace at which symbols _outside_ of the workspace are referenced. That is:
+
+- Excluding any `ReferenceInformation.uri` which is located within the workspace.
+- Excluding any `ReferenceInformation.name` which is unexported or private (i.e. according to language semantics).
+- Excluding any `ReferenceInformation.location` which is located in vendored code (e.g. `vendor/...` for Go, `node_modules/...` for JS, .tgz NPM packages, or .jar files for Java).
+
+
+_Request_
+* method: 'workspace/reference'
+* params: `WorkspaceReferenceParams` defined as follows:
+```typescript
+/**
+ * The parameters of a Workspace Reference Request.
+ */
+interface WorkspaceReferenceParams {
+	/**
+	 * A limit on the number of returned results.
+	 */
+	limit: number
+}
+```
+
+_Response_:
+* result: `ReferenceInformation[]` defined as follows:
+```typescript
+/**
+ * Represents information about a reference to programming constructs like
+ * variables, classes, interfaces etc.
+ */
+interface ReferenceInformation {
+	/**
+	 * The location in the workspace where the `symbol` has been referenced.
+	 */
+	location: Location;
+
+	/**
+	 * The name of the symbol that is being referenced. For example:
+	 *
+	 * 	(Go) "ServeHTTP"
+	 *  (JS) "render"
+	 *
+	 */
+	name?: string;
+
+	/**
+	 * The name of the symbol containing the symbol that is being referenced.
+	 * For example:
+	 *
+	 * 	(Go) "Router"
+	 * 	(JS) "ReactMount"
+	 *
+	 */
+	containerName?: string;
+
+	/**
+	 * The file URI of the symbol that is being referenced. It may be either
+	 * a URI to the directory that contains the symbol OR the file that
+	 * contains the symbol, so long as the tuple (name, containerName, uri)
+	 * uniquely identifies the symbol.
+	 *
+	 * If `name` and `containerName` are not present, it indicates that this
+	 * URI location has been referenced but not a particular symbol. For
+	 * example, a Go import statement or a JS `require()` call, etc. which
+	 * references ('imports' or 'includes') a dependency but not a specific
+	 * symbol.
+	 */
+	uri: string;
+}
+```
+* error: code and message set in case an exception happens during the workspace reference request.
+
+## workspace/symbol query filters
+
+As part of this extension, we also require that language servers support two `workspace/symbol` query filters (specified via the `WorkspaceSymbolParams.query` field). These filters are:
+
+- `is:exported`
+  - When this filter is present in the query, only 'exported' symbols (according to language semantics) should be returned.
+- `dir:<directory>`
```

Comment:
This reminds me of the `base` param in the `workspace/files` proposal, which is also used to filter the results to only a specific directory. Instead of a query DSL, we could do all these fancy filters with additional params, which means the LS doesn't have to do any fancy parsing and it is probably also easier for Sourcegraph to implement. If this request is exposed to the user directly somewhere (and we should assume this for all requests) from a UI standpoint this is the difference of a user having to remember the DSL and typing it into a search box, or having checkboxes and extra input fields for these special filters. As this is a workspace request, in VS Code for example I could imagine the latter more (similar how simple workspace file search works, in a side panel).

If a DSL is used, how are characters like space or colon escaped, as they have special meaning?